File: sundarr/app/schemas/transfer.py
from typing import Literal

from pydantic import BaseModel, Field

TransferStatus = Literal[
    "pending",
    "staging_to_cloud",
    "cloud_ready",
    "downloading",
    "verifying",
    "renaming",
    "cleaning_cloud",
    "cleaning_source",
    "completed",
    "failed",
    "cancelled",
    "paused",
]


class TransferCreateRequest(BaseModel):
    link_id: str = Field(min_length=1)
    mode: str = "copy"
    target_type: Literal["smb"] = "smb"
    target_library: str | None = None
    target_path: str = Field(min_length=1)


class TransferResponse(BaseModel):
    id: str
    resource_id: str | None = None
    link_id: str | None = None
    status: TransferStatus
    mode: str
    cloud_staging_path: str | None = None
    target_type: str
    target_library: str | None = None
    target_path: str
    source_type: str | None = None
    source_path: str | None = None
    sync_seen_file_id: str | None = None
    total_bytes: int
    done_bytes: int
    speed_bytes_per_sec: int = 0
    progress: float = 0
    current_file: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    retryable: bool | None = None
    retry_count: int
    created_at: str | None = None
    updated_at: str | None = None


class TransferListResponse(BaseModel):
    count: int
    page: int = 1
    page_size: int = 20
    results: list[TransferResponse]


class TransferLogResponse(BaseModel):
    id: str
    task_id: str
    level: str
    event: str
    message: str | None = None
    data: dict | None = None
    created_at: str

File: web/src/types/transfer.ts
export type TransferStatus =
  | 'pending'
  | 'staging_to_cloud'
  | 'cloud_ready'
  | 'downloading'
  | 'verifying'
  | 'renaming'
  | 'cleaning_cloud'
  | 'cleaning_source'
  | 'completed'
  | 'failed'
  | 'cancelled'
  | 'paused'

export type TransferResponse = {
  id: string
  resource_id: string | null
  link_id: string | null
  status: TransferStatus
  mode: string
  cloud_staging_path: string | null
  target_type: string
  target_library: string | null
  target_path: string
  source_type: string | null
  source_path: string | null
  sync_seen_file_id: string | null
  total_bytes: number
  done_bytes: number
  speed_bytes_per_sec: number
  progress: number
  current_file: string | null
  error_code: string | null
  error_message: string | null
  retryable: boolean | null
  retry_count: number
  created_at: string | null
  updated_at: string | null
}

export type TransferLogResponse = {
  id: string
  task_id: string
  level: string
  event: string
  message: string | null
  data: Record<string, unknown> | null
  created_at: string
}

export type TransferListResponse = {
  count: number
  page: number
  page_size: number
  results: TransferResponse[]
}
